nexus/operation-ids.md

```diff
@@ -0,0 +1,172 @@
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+
+- [Nexus RPCs - short recap](#nexus-rpcs---short-recap)
+- [Operation identifiers](#operation-identifiers)
+  - [Caller generated identifiers](#caller-generated-identifiers)
+    - [Handler addressability](#handler-addressability)
+      - [Mapping from client to handler identifiers](#mapping-from-client-to-handler-identifiers)
+  - [Handler generated identifiers](#handler-generated-identifiers)
+    - [Cancel before started](#cancel-before-started)
+  - [Comparing the approaches](#comparing-the-approaches)
+    - [Handler Id qualification](#handler-id-qualification)
+    - [Ids as inputs](#ids-as-inputs)
+    - [Caller friendliness](#caller-friendliness)
+    - [Data sensitivity](#data-sensitivity)
+- [Proposal](#proposal)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Nexus RPCs - short recap
+
+At the fundamental level, the Nexus protocol is a synchronous RPC protocol. Asynchronous operations are modelled on top
+of a set of pre-defined RPCs.
+
+A Nexus **caller** calls a **handler**. The handler may respond inline or return a reference for a future, asynchronous
+operation. These asynchronous operations can be cancelled and checked for their outcomes and current state. The caller
+can also specify a callback URL, which the handler uses to asynchronously deliver the result of an operation's when it
+is ready.
+
+## Operation identifiers
+
+Since an asynchronous operation is modelled on top of synchronous RPCs, it must be addressable (identifiable) by both
+the caller and the handler. There are two approaches for generating identifiers, caller side and handler side.
+
+### Caller generated identifiers
+
+With this approach, the caller provides an Id to each of the pre-defined RPCs.
+
+The handler _should_ qualify an identifier when dealing with untrusted sources and multi-tenant applications.
+
```

Comment:
This is not enough when you are trying to share access with others, e.g.,  give the id to someone else so that they can GetResult(id) safely. In those cases you want the identifiers to be **unguessable**, so that they can behave as an object capability. The client is hopeless for that, e.g., it may not have enough entropy in the RNG, so if sharing is important we need handler ids...

Reply:
> In those cases you want the identifiers to be unguessable

I am not sure we want that though. Often letting a user choose an identifier that is meaningful to them so they can reuse that is ideal.

Reply:
Unguessable helps with fine-grained access control. If they are unguessable you can use them as a capability, so if you have it, you also have the right to see the result, and that simplifies safe sharing.

But you are right, you cannot expect clients to directly use these ids, they are not user-friendly, you need to provide local, persistent aliases to access them, so that they can choose names meaningful to them.

Reply:
Added a section in the proposal (will push soon)